src/eko/output/struct.py

```diff
@@ -0,0 +1,502 @@
+    limitations in memory are far more strict than on disk.
+    Since manually managing, for each application, the burden of off-loading
+    part of the operator might be hard and occasionally not possible (without a
+    clear picture of the internals), the library itself offers this facility.
+
+    In particular, the data format on disk has a complete specification, and
+    can hold a full operator independently of the loading procedure.
+    In order to accomplish the former goal, the remaining task of partial
+    loading is done by this class (for the Python library, other
+    implementations are possible and encouraged).
+
+    For this reason, a core component of an :cls:`EKO` object is a path,
+    referring to the location on disk of the corresponding operator.
+    Any :cls:`EKO` has an associated path:
+
+    - for the computed object, it corresponds to the path where the actual
+    result of the computation is already saved
+    - for a new object, it is the path at which any result of final or
+    intermediate computation is stored, as soon as it is produced
+
+    The computation can be stopped at any time, without the loss of any of the
+    intermediate results.
+
+    Attributes
+    ----------
+    path : pathlib.Path
+        path on disk, to which this object is linked (and for which it is
+        essentially an interface)
+    Q02 : float
+        inital scale
+    xgrid : interpolation.XGrid
+        momentum fraction internal grid
+    pids : np.ndarray
+        array of integers, corresponding to internal PIDs
+    configs : Configs
+        specific configuration to be used during the calculation of these
+        operators
+    rotations : Rotations
+        manipulation information, describing the current status of the EKO (e.g.
+        `inputgrid` and `targetgrid`)
+    debug : Debug
+        debug configurations
+    version : str
+        library version used to create the corresponding file
+    data_version : str
+        specs version, to which the file adheres
+
+    """
+
+    # operators cache, contains the Q2 grid information
+    _operators: Dict[float, Optional[Operator]]
+    # public attributes
+    # -----------------
+    # mandatory, identifying features
+    path: pathlib.Path
+    Q02: float
+    xgrid: interpolation.XGrid
+    pids: np.ndarray
+    # collections
+    configs: Configs
+    rotations: Rotations
+    debug: Debug
+    # tagging information
+    version: str = vmod.__version__
+    data_version: str = vmod.__data_version__
+
+    def __post_init__(self):
+        if self.path.suffix != ".tar":
+            raise ValueError("Not a valid path for an EKO")
+        if not tarfile.is_tarfile(self.path):
+            raise ValueError("EKO: the corresponding file is not a valid tar archive")
+
+    def __getitem__(self, q2: float):
+        # TODO: autoload
+        op = self._operators[q2]
+        if op is not None:
+            return op
+
+        start = f"operators/{q2:8.2f}"
```

Comment:
delegate to a method `def op_name(q2)` (since it is repeated and crucial) - or actual you can also just expose as (module-wide) global format string

Reply:
hm? also f-strings can be constants - simply without the f and then you "call" them with `var.format(a=b)` (the f-syntax is just sugar)

Reply:
Solved in 1be6fa4e8faedfa49065fbacfa5adcb86b3554d0